numba/np/npyimpl.py

```diff
@@ -315,47 +336,58 @@ def numpy_ufunc_kernel(context, builder, sig, args, kernel_class,
     # builder - the code emitter
     # sig - signature of the ufunc
     # args - the args to the ufunc
+    # ufunc - the ufunc itself
     # kernel_class -  a code generating subclass of _Kernel that provides
-    # explicit_output - if the output was explicit in the call
-    #                   (ie: np.add(x,y,r))
 
     arguments = [_prepare_argument(context, builder, arg, tyarg)
                  for arg, tyarg in zip(args, sig.args)]
-    if not explicit_output:
-        ret_ty = sig.return_type
-        if isinstance(ret_ty, types.ArrayCompatible):
-            output = _build_array(context, builder, ret_ty, sig.args, arguments)
-        else:
-            output = _prepare_argument(
-                context, builder,
-                lc.Constant.null(context.get_value_type(ret_ty)), ret_ty)
-        arguments.append(output)
-    elif context.enable_nrt:
-        # Incref the output
-        context.nrt.incref(builder, sig.return_type, args[-1])
-
-    inputs = arguments[0:-1]
-    output = arguments[-1]
-
-    outer_sig = [a.base_type for a in arguments]
-    #signature expects return type first, while we have it last:
-    outer_sig = outer_sig[-1:] + outer_sig[:-1]
-    outer_sig = typing.signature(*outer_sig)
+
+    if len(arguments) < ufunc.nin:
+        raise RuntimeError(
+            "Not enough inputs to {}, expected {} got {}"
+            .format(ufunc.__name__, ufunc.nin, len(arguments)))
+
+    for out_i, ret_ty in enumerate(_unpack_output_types(ufunc, sig)):
+        if ufunc.nin + out_i >= len(arguments):
+            # this out argument is not provided
+            if isinstance(ret_ty, types.ArrayCompatible):
+                output = _build_array(context, builder, ret_ty, sig.args, arguments)
+            else:
+                output = _prepare_argument(
+                    context, builder,
+                    lc.Constant.null(context.get_value_type(ret_ty)), ret_ty)
+            arguments.append(output)
+        elif context.enable_nrt:
+            # Incref the output
+            context.nrt.incref(builder, ret_ty, args[ufunc.nin + out_i])
+
+    inputs = arguments[:ufunc.nin]
+    outputs = arguments[ufunc.nin:]
+    assert len(outputs) == ufunc.nout
+
+    outer_sig = _ufunc_loop_sig(
+        [a.base_type for a in outputs],
+        [a.base_type for a in inputs]
+    )
     kernel = kernel_class(context, builder, outer_sig)
     intpty = context.get_value_type(types.intp)
 
     indices = [inp.create_iter_indices() for inp in inputs]
 
-    loopshape = output.shape
+    # assume outputs are all the same size
```

Comment:
Perhaps add something to catch if they aren't?

Reply:
Numba is already broken when it comes to checking the shape of output arrays, so I'd prefer to declare this out of scope. Without this PR:
```
In [11]: @numba.njit
    ...: def neg(a, b):
    ...:     return np.negative(a, b)
    ...:

In [16]: neg(np.ones(2, int), np.zeros((2, 4), int))
Out[16]:
array([[         -1,          -1, -1409738722,  2013264680],
       [         -1,          -1, -1409738722,  2013264680]])
```

Reply:
At any rate, numpy does not allow them to be different sizes.

Reply:
ok, fair point, something to address another time.